# Data Cleaning-Printer

The reviews of different Laptops, Smart Phones, Headphones, Smart Watches, DSLR (Professional Cameras), Printers, Monitors, Home Theaters, Routers from Flipkart.
The data preprocessing will be done on each of the dataset seaprately. First we will filter out the data to get equal number of reviews for each rating. The heading of the review is also extracted as the heading also can help in determining the rating.

In [18]:
#import the dataset
import pandas as pd
import numpy as np
pd.set_option('Display.max_columns',None)
pd.set_option('Display.max_rows',None)

df=pd.read_csv("Printer Rating.csv")
df.head()

Unnamed: 0  Rating           Heading  \
0           0     5.0  Perfect product!   
1           1     5.0         Excellent   
2           2     5.0         Brilliant   
3           3     NaN               NaN   
4           4     NaN               NaN   

                                              Review  Product  
0  Very nice printer, is worth the purchase. Idea...  Printer  
1  great !! easy to use ,, budget friendly,, good...  Printer  
2  nice product\nI am a student I ordered this fo...  Printer  
3                                                NaN  Printer  
4                                                NaN  Printer

#### Observations:
* The feature unnamed is index. Hence we can drop this feature.

In [19]:
df.drop(['Unnamed: 0'], axis=1, inplace=True)
df.head()

Rating           Heading  \
0     5.0  Perfect product!   
1     5.0         Excellent   
2     5.0         Brilliant   
3     NaN               NaN   
4     NaN               NaN   

                                              Review  Product  
0  Very nice printer, is worth the purchase. Idea...  Printer  
1  great !! easy to use ,, budget friendly,, good...  Printer  
2  nice product\nI am a student I ordered this fo...  Printer  
3                                                NaN  Printer  
4                                                NaN  Printer

### Exploratory Data Analysis

In [20]:
#check the dimensions of the data (Headphone)
df.shape

(14939, 4)

* The dataset has 14939 rows and 4 columns
* The dataset has 1 label - 'Rating' and 3 features

In [21]:
#check the names of columns in dataset
df.columns

Index(['Rating', 'Heading', 'Review', 'Product'], dtype='object')

In [22]:
#check the datatype of each feature
df.dtypes

Rating     float64
Heading     object
Review      object
Product     object
dtype: object

#### Observations:
   * All the feratures are of "object" data type.

In [23]:
#checking if there are any null values in the dataset
df.isna().sum()

Rating     456
Heading    429
Review     427
Product      0
dtype: int64

In [24]:
df[df.isna()]

Rating Heading Review Product
0         NaN     NaN    NaN     NaN
1         NaN     NaN    NaN     NaN
2         NaN     NaN    NaN     NaN
3         NaN     NaN    NaN     NaN
4         NaN     NaN    NaN     NaN
5         NaN     NaN    NaN     NaN
6         NaN     NaN    NaN     NaN
7         NaN     NaN    NaN     NaN
8         NaN     NaN    NaN     NaN
9         NaN     NaN    NaN     NaN
10        NaN     NaN    NaN     NaN
11        NaN     NaN    NaN     NaN
12        NaN     NaN    NaN     NaN
13        NaN     NaN    NaN     NaN
14        NaN     NaN    NaN     NaN
15        NaN     NaN    NaN     NaN
16        NaN     NaN    NaN     NaN
17        NaN     NaN    NaN     NaN
18        NaN     NaN    NaN     NaN
19        NaN     NaN    NaN     NaN
20        NaN     NaN    NaN     NaN
21        NaN     NaN    NaN     NaN
22        NaN     NaN    NaN     NaN
23        NaN     NaN    NaN     NaN
24        NaN     NaN    NaN     NaN
25        NaN     NaN    NaN     NaN
26        NaN     NaN    NaN     NaN
27        NaN     NaN    NaN     NaN
28        NaN     NaN    NaN     NaN
29        NaN     NaN    NaN     NaN
30        NaN     NaN    NaN     NaN
31        NaN     NaN    NaN     NaN
32        NaN     NaN    NaN     NaN
33        NaN     NaN    NaN     NaN
34        NaN     NaN    NaN     NaN
35        NaN     NaN    NaN     NaN
36        NaN     NaN    NaN     NaN
37        NaN     NaN    NaN     NaN
38        NaN     NaN    NaN     NaN
39        NaN     NaN    NaN     NaN
40        NaN     NaN    NaN     NaN
41        NaN     NaN    NaN     NaN
42        NaN     NaN    NaN     NaN
43        NaN     NaN    NaN     NaN
44        NaN     NaN    NaN     NaN
45        NaN     NaN    NaN     NaN
46        NaN     NaN    NaN     NaN
47        NaN     NaN    NaN     NaN
48        NaN     NaN    NaN     NaN
49        NaN     NaN    NaN     NaN
50        NaN     NaN    NaN     NaN
51        NaN     NaN    NaN     NaN
52        NaN     NaN    NaN     NaN
53        NaN     NaN    NaN     NaN
54        NaN     NaN    NaN     NaN
55        NaN     NaN    NaN     NaN
56        NaN     NaN    NaN     NaN
57        NaN     NaN    NaN     NaN
58        NaN     NaN    NaN     NaN
59        NaN     NaN    NaN     NaN
60        NaN     NaN    NaN     NaN
61        NaN     NaN    NaN     NaN
62        NaN     NaN    NaN     NaN
63        NaN     NaN    NaN     NaN
64        NaN     NaN    NaN     NaN
65        NaN     NaN    NaN     NaN
66        NaN     NaN    NaN     NaN
67        NaN     NaN    NaN     NaN
68        NaN     NaN    NaN     NaN
69        NaN     NaN    NaN     NaN
70        NaN     NaN    NaN     NaN
71        NaN     NaN    NaN     NaN
72        NaN     NaN    NaN     NaN
73        NaN     NaN    NaN     NaN
74        NaN     NaN    NaN     NaN
75        NaN     NaN    NaN     NaN
76        NaN     NaN    NaN     NaN
77        NaN     NaN    NaN     NaN
78        NaN     NaN    NaN     NaN
79        NaN     NaN    NaN     NaN
80        NaN     NaN    NaN     NaN
81        NaN     NaN    NaN     NaN
82        NaN     NaN    NaN     NaN
83        NaN     NaN    NaN     NaN
84        NaN     NaN    NaN     NaN
85        NaN     NaN    NaN     NaN
86        NaN     NaN    NaN     NaN
87        NaN     NaN    NaN     NaN
88        NaN     NaN    NaN     NaN
89        NaN     NaN    NaN     NaN
90        NaN     NaN    NaN     NaN
91        NaN     NaN    NaN     NaN
92        NaN     NaN    NaN     NaN
93        NaN     NaN    NaN     NaN
94        NaN     NaN    NaN     NaN
95        NaN     NaN    NaN     NaN
96        NaN     NaN    NaN     NaN
97        NaN     NaN    NaN     NaN
98        NaN     NaN    NaN     NaN
99        NaN     NaN    NaN     NaN
100       NaN     NaN    NaN     NaN
101       NaN     NaN    NaN     NaN
102       NaN     NaN    NaN     NaN
103       NaN     NaN    NaN     NaN
104       NaN     NaN    NaN     NaN
105       NaN     NaN    NaN     NaN
106       NaN     NaN    NaN     NaN
107       N

In [25]:
#dropping all numm values
df.dropna(inplace=True)

In [26]:
#cross checking null values
df.isna().sum()

Rating     0
Heading    0
Review     0
Product    0
dtype: int64

In [27]:
df.shape

(14481, 4)

#### Observations:
* There are 14481 rows in the dataset

In [28]:
df['Rating'].value_counts()

5.0    6436
1.0    3391
4.0    2544
3.0    1333
2.0     777
Name: Rating, dtype: int64

#### Observations:
* The labels are inbalanced. 
* The efficiency of review classifier will be better when we have equal or near equal number of reviews for each rating

#### Action:
* We will have as many number of reviews for each rating as there are in the 1 rating (least rating)
* The excess reviews dropped.
* Before dropping the excess reviews we will first check the length of reviews and drop long reviews as it will use up more space.

In [29]:
#check info of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14481 entries, 0 to 14938
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Rating   14481 non-null  float64
 1   Heading  14481 non-null  object 
 2   Review   14481 non-null  object 
 3   Product  14481 non-null  object 
dtypes: float64(1), object(3)
memory usage: 565.7+ KB


#### Observations
   * The info() method thus returns the data type as well as the non-null values and memory usage.
   * Out of the total of 4 columns 3 columns are "object" type while rest of them are float datatype.

In [30]:
#check number of unique values in each class;
df.nunique()

Rating        5
Heading     718
Review     5580
Product       1
dtype: int64

#### Observations:
* The label rating has 5 unique values: 1, 2, 3, 4, 5
* The headings can be duplicate as it is kind of summary of the review.
* The reviews should be unique. Hence, we will drop the duplicate reviews to avoid over-fitting.

In [31]:
df.drop_duplicates(subset='Review', inplace=True)

In [32]:
#cross checking for diplicacy of reviews
print(df.shape)
print(df.nunique())

(5580, 4)
Rating        5
Heading     697
Review     5580
Product       1
dtype: int64


In [33]:
#checking the length of review
df['Review_word_counter']=df['Review'].str.strip().str.len()
df.head()

Rating                 Heading  \
0      5.0        Perfect product!   
1      5.0               Excellent   
2      5.0               Brilliant   
9      1.0  Not recommended at all   
10     5.0            nice product   

                                               Review  Product  \
0   Very nice printer, is worth the purchase. Idea...  Printer   
1   great !! easy to use ,, budget friendly,, good...  Printer   
2   nice product\nI am a student I ordered this fo...  Printer   
9   printer is very good but it's cartage is not g...  Printer   
10  its a good printer, but i found it little slow...  Printer   

    Review_word_counter  
0                   510  
1                   110  
2                   201  
9                   202  
10                  119

The maximum number of reviews are for 5 star rating. We will sort the rating in ascending order and length of review. So, that we drop excess number of reviews for 5 star rating and get near balanced dataset.

In [34]:
sorted_df=df.sort_values(by=['Rating','Review_word_counter'])
sorted_df.head()

Rating                 Heading Review  Product  Review_word_counter
566       1.0         Waste of money!    Bad  Printer                    3
4036      1.0  Worst experience ever!    pad  Printer                    3
5298      1.0                 Awesome    top  Printer                    3
14259     1.0                    Good    NIC  Printer                    3
1363      1.0               Worthless   Poor  Printer                    4

In [35]:
sorted_df.head(100)

Rating                      Heading         Review  Product  \
566       1.0              Waste of money!            Bad  Printer   
4036      1.0       Worst experience ever!            pad  Printer   
5298      1.0                      Awesome            top  Printer   
14259     1.0                         Good            NIC  Printer   
1363      1.0                    Worthless           Poor  Printer   
2686      1.0       Not recommended at all           wast  Printer   
1167      1.0            Absolute rubbish!          Worst  Printer   
1653      1.0       Worst experience ever!          Wrost  Printer   
2461      1.0    Did not meet expectations          waste  Printer   
3218      1.0                    Very poor          Canon  Printer   
6556      1.0             Met Expectations          Scrap  Printer   
14913     1.0       Not recommended at all          worst  Printer   
388       1.0                     Horrible         to gud  Printer   
507       1.0    Did not meet expectations         Owsame  Printer   
668       1.0         Utterly Disappointed         No bad  Printer   
961       1.0       Not recommended at all         Normal  Printer   
1112      1.0       Worst experience ever!         So bad  Printer   
2703      1.0            Absolute rubbish!         Simple  Printer   
4182      1.0       Worst experience ever!         Nt bad  Printer   
4292      1.0              Waste of money!         Shoddy  Printer   
4665      1.0                    Hated it!         To bad  Printer   
6170      1.0       Not recommended at all         Poor..  Printer   
6505      1.0    Did not meet expectations         Bakwas  Printer   
1108      1.0               Unsatisfactory        Damaged  Printer   
1196      1.0              Waste of money!        No Good  Printer   
1474      1.0       Don't waste your money        Problem  Printer   
2223      1.0            Absolute rubbish!        So slow  Printer   
2290      1.0             Terrible product        No like  Printer   
2832      1.0                    Worthless        too bad  Printer   
348       1.0       Worst experience ever!       VERY BAD  Printer   
428       1.0                    Worthless       Excelent  Printer   
485       1.0       Worst experience ever!       Very bad  Printer   
1045      1.0                     Horrible       nice one  Printer   
1459      1.0       Worst experience ever!       Very bed  Printer   
2528      1.0              Useless product       Very Bad  Printer   
3063      1.0       Not recommended at all       problem.  Printer   
4336      1.0                Great product       just gud  Printer   
4343      1.0              Useless product       not nice  Printer   
5240      1.0       Don't waste your money       Dont buy  Printer   
5530      1.0    Did not meet expectations       Pathetic  Printer   
5896      1.0              Useless product       Not like  Printer   
7221      1.0                     Horrible       Hopeless  Printer   
434       1.0         Utterly Disappointed      Don't buy  Printer   
667       1.0       Worst experience ever!      don't buy  Printer   
670       1.0       Worst experience ever!      3rd class  Printer   
784       1.0       Don't waste your money      Very poor  Printer   
1075      1.0              Waste of money!      Good unit  Printer   
3008      1.0              Useless product      Not good.  Printer   
3215      1.0                    Very poor      disapoint  Printer   
4260      1.0         Utterly Disappointed      Only good  Printer   
4312      1.0       Not recommended at all      verry bad  Printer   
4556      1.0                     Horrible      bad items  Printer   
4816      1.0                    Hated it!      Not happy  Printer   
5397      1.0     waste printer i ever saw      dont buy.  Printer   
512       1.0       Not recommended at all     All in one  Printer   
556       1.0            Absolute rubbish!     Worst item  Printer   
1392      1.0    

In [36]:
sorted_df['Rating'].value_counts()

5.0    2301
1.0    1463
4.0     951
3.0     525
2.0     340
Name: Rating, dtype: int64

The least number of reviews are for rating '2'. We will make the number of reviews for rating '4' and '5' equal to the number of reviews for rating '1'.

In [37]:
df_1=sorted_df[sorted_df['Rating']==1]
df_2=sorted_df[sorted_df['Rating']==2]
df_3=sorted_df[sorted_df['Rating']==3]
df_4=sorted_df[sorted_df['Rating']==4]
df_5=sorted_df[sorted_df['Rating']==5]

In [38]:
df_printer=pd.concat([df_1[:951],df_2,df_3,df_4,df_5[:951]])
df_printer.head()

Rating                 Heading Review  Product  Review_word_counter
566       1.0         Waste of money!    Bad  Printer                    3
4036      1.0  Worst experience ever!    pad  Printer                    3
5298      1.0                 Awesome    top  Printer                    3
14259     1.0                    Good    NIC  Printer                    3
1363      1.0               Worthless   Poor  Printer                    4

In [39]:
df_printer.shape

(3718, 5)

In [40]:
#Reshuffling and reindexing the data
from sklearn.utils import shuffle
df_printer=shuffle(df_printer)
df_printer.reset_index(inplace=True,drop=True)
df_printer.head(100)

Rating                                            Heading  \
0      2.0                                 BAKWAS, HP is best   
1      1.0                                          Very poor   
2      5.0                                     Simply awesome   
3      4.0                                          Wonderful   
4      4.0                                        Pretty good   
5      3.0  not so good but overall useful if it is workin...   
6      5.0                                        Nic product   
7      4.0                                         Delightful   
8      1.0                             Worst experience ever!   
9      4.0                                    Value-for-money   
10     4.0                               Good quality product   
11     4.0                                          Very Good   
12     3.0                                     Decent product   
13     2.0                                           Not good   
14     5.0                                          Wonderful   
15     4.0                                        Really Nice   
16     4.0                                       Nice product   
17     3.0                                               Nice   
18     3.0                                     Decent product   
19     5.0                              Mind-blowing purchase   
20     5.0                                           Terrific   
21     1.0                                    Useless product   
22     5.0                                          Fabulous!   
23     4.0                                        Really Nice   
24     1.0                          Did not meet expectations   
25     1.0                             Worst experience ever!   
26     4.0  Good for price of 2k. But ink given is very less.   
27     5.0                                          Fabulous!   
28     4.0                                        Pretty good   
29     4.0                                       Good Product   
30     4.0                                    Worth the money   
31     5.0                                  Worth every penny   
32     4.0                                     paper required   
33     4.0                                       Nice product   
34     1.0                                     Unsatisfactory   
35     4.0                    Good product but has its warts.   
36     5.0                                 Highly recommended   
37     5.0                                           Terrific   
38     5.0                                Best in the market!   
39     5.0                                          Brilliant   
40     4.0                                        Pretty good   
41     4.0                                          Very Good   
42     2.0                                        Bad quality   
43     4.0                               Good quality product   
44     2.0                                         Five Stars   
45     5.0                                            Awesome   
46     1.0                                   Terrible product   
47     5.0                                      Great product   
48     5.0                                  Terrific purchase   
49     4.0                                    Worth the money   
50     5.0                                  Terrific purchase   
51     5.0                                               Good   
52     4.0                                             Gainer   
53     5.0                                             Super!   
54     5.0                                     Classy product   
55     5.0                                          Brilliant   
56     1.0                                    Waste of money!   
57     5.0                                Best in the market!   
58     3.0                                       Does the job   
59     5.0                                          Just wow!   
60     4.0                             

In [41]:
df_printer.drop(['Review_word_counter'],axis=1,inplace=True)
df_printer.head()

Rating             Heading  \
0     2.0  BAKWAS, HP is best   
1     1.0           Very poor   
2     5.0      Simply awesome   
3     4.0           Wonderful   
4     4.0         Pretty good   

                                              Review  Product  
0  just prints....but print quality is not at all...  Printer  
1  Very bad.... After refilling its not working.....  Printer  
2                                       invoice sand  Printer  
3  Awesome printer, for school, college and offic...  Printer  
4  Functioning good but not he cartridge get empt...  Printer

In [42]:
df_printer.to_csv('Printer Rating_cleaned.csv')

In [43]:
df_printer.shape

(3718, 4)